File: languages/go/client/entity/loader.go
package entity

import "github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"

type Loader interface {
	ToProto() *proto.Loader
}

type LoaderArgs struct {
	hash string
}

func (e LoaderArgs) ToProto() *proto.LoaderArgs {
	return &proto.LoaderArgs{
		Hash: e.hash,
	}
}

type HostedLoader struct {
	Type proto.DataAvailabilityType
	Args LoaderArgs
}

func NewHostedLoader(hash string) HostedLoader {
	return HostedLoader{
		Type: proto.DataAvailabilityType_HOSTED,
		Args: LoaderArgs{
			hash,
		},
	}
}

func (e HostedLoader) ToProto() *proto.Loader {
	return &proto.Loader{
		Type: e.Type,
		Args: e.Args.ToProto(),
	}
}

type IpfsLoader struct {
	Type proto.DataAvailabilityType
	Args LoaderArgs
}

func NewIpfsLoader(hash string) IpfsLoader {
	return IpfsLoader{
		Type: proto.DataAvailabilityType_IPFS,
		Args: LoaderArgs{
			hash,
		},
	}
}

func (e IpfsLoader) ToProto() *proto.Loader {
	return &proto.Loader{
		Type: e.Type,
		Args: e.Args.ToProto(),
	}
}

File: languages/python/bloock/entity/loader.py
from abc import abstractmethod
import bloock._bridge.proto.availability_entities_pb2 as proto


class LoaderArgs:
    def __init__(self, id: str) -> None:
        self.id = id

    def to_proto(self) -> proto.LoaderArgs:
        return proto.LoaderArgs(id=self.id)


class Loader:
    def __init__(
        self, type: proto.DataAvailabilityType.ValueType, args: LoaderArgs
    ) -> None:
        self.type = type
        self.args = args

    @abstractmethod
    def to_proto(self):
        raise NotImplementedError


class HostedLoader(Loader):
    def __init__(self, hash: str) -> None:
        super().__init__(type=proto.HOSTED, args=LoaderArgs(hash))

    def to_proto(self) -> proto.Loader:
        return proto.Loader(type=self.type, args=self.args.to_proto())


class IpfsLoader(Loader):
    def __init__(self, hash: str) -> None:
        super().__init__(type=proto.IPFS, args=LoaderArgs(hash))

    def to_proto(self) -> proto.Loader:
        return proto.Loader(type=self.type, args=self.args.to_proto())
